src/Symfony/Bundle/FrameworkBundle/Test/WebTestCase.php

```diff
@@ -58,4 +62,62 @@ protected static function createClient(array $options = [], array $server = [])
 
         return self::getClient($client);
     }
+
+    /**
+     * Logouts a user.
+     */
+    public function logout(KernelBrowser $browser, string $firewallName = 'main'): void
+    {
+        $context = $this->getFirewallContext($browser, $firewallName);
+
+        $browser->getContainer()->get('session')->remove('_security_'.$context);
+    }
+
+    public function login(
+        KernelBrowser $browser,
+        $user,
+        array $roles = [],
+        string $firewallName = 'main'
+    ): TokenInterface {
+        if (!class_exists(UsernamePasswordToken::class)) {
+            throw new \RuntimeException('You must install the "symfony/security-core" component to use this feature.');
+        }
+
+        if ($user instanceof UserInterface) {
+            $roles = array_merge($user->getRoles(), $roles);
+        }
+        $token = $this->getLoginToken($browser, $user, $roles, $firewallName);
+
+        $this->authenticateToken($browser, $token, $firewallName);
+
+        return $token;
+    }
+
+    protected function getLoginToken(KernelBrowser $browser, $user, array $roles = [], string $firewallName = 'main'): TokenInterface
+    {
+        return new UsernamePasswordToken($user, null, $firewallName, $roles);
+    }
+
+    protected function authenticateToken(KernelBrowser $browser, TokenInterface $token, string $firewallName = 'main'): void
+    {
+        $context = $this->getFirewallContext($browser, $firewallName);
+
+        $session = $browser->getContainer()->get('session');
+        $session->set('_security_'.$context, serialize($token));
+        $session->save();
+
+        $cookie = new Cookie($session->getName(), $session->getId());
+        $browser->getCookieJar()->set($cookie);
+    }
+
+    private function getFirewallContext(KernelBrowser $browser, string $firewallName = 'main'): string
+    {
+        $config = sprintf('security.firewall.map.config.%s', $firewallName);
+
+        if (!$browser->getContainer()->has($config)) {
```

Comment:
I just tried it and this will always return `false`, since the `security.firewall.map.config.*` services are private.

I have a question: is there a way to make the `$browser->getContainer()` return the same container as `static::$container` so we can retrieve private services? I don't know the impact in performances or anything since I don't really understand how it is managed internally. @nicolas-grekas, any idea?

This would be transparent for the end user, and still make sure that anyone could retrieve any service from the client when the container is an instance of the `TestContainer` (and of course, there will be an additionnal check here)